### PR TITLE
[Merged by Bors] - feat: Order on the localization

### PR DESCRIPTION
**Pull request by YaelDillies**
Match https://github.com/leanprover-community/mathlib/pull/18724 and https://github.com/leanprover-community/mathlib/pull/18846

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

[`group_theory.monoid_localization`@`1f0096e6caa61e9c849ec2adbd227e960e9dff58`..`13b8e258f14bffb5def542aa78b803b0b80541aa`](https://leanprover-community.github.io/mathlib-port-status/file/group_theory/monoid_localization?range=1f0096e6caa61e9c849ec2adbd227e960e9dff58..13b8e258f14bffb5def542aa78b803b0b80541aa)

### CODE REVIEW

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 1]
```diff
@@ -1866,3 +1888,154 @@ end LocalizationWithZeroMap
...
+instance [LinearOrderedCancelCommMonoid α] {s : Submonoid α} :
+    LinearOrderedCancelCommMonoid (Localization s) :=
+  { Localization.orderedCancelCommMonoid with
+    le_total := fun a b =>
+      Localization.induction_on₂ a b fun _ _ =>
+        by
+        simp_rw [mk_le_mk]
+        exact le_total _ _
+    decidable_le := @Localization.decidableLe α _ _ LE.le.decidable
+    decidable_lt := @Localization.decidableLt α _ _ LT.lt.decidable
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:2038]:
```suggestion
    decidable_lt := @Localization.decidableLt α _ _ LT.lt.decidable  -- porting note: was wrong in mathlib3
```
or something, assuming you don't want to wait for a fix in mathlib3 first

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 2]
```diff
@@ -301,6 +312,17 @@ def rec {p : Localization S → Sort u} (f : ∀ (a : M) (b : S), p (mk a b))
 #align localization.rec Localization.rec
 #align add_localization.rec addLocalization.rec
 
+/-- Copy of `quotient.rec_on_subsingleton₂` for `localization` -/
+@[to_additive (attr := elab_as_elim) "Copy of `quotient.rec_on_subsingleton₂` for
+`add_localization`"]
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:317]: These names should be mathlib4-ified

File: Mathlib/GroupTheory/MonoidLocalization.lean [hunk 3]
```diff
@@ -1866,3 +1888,158 @@ end LocalizationWithZeroMap
...
 
 end CommMonoidWithZero
+
+namespace Localization
+
+variable {α : Type _} [CancelCommMonoid α] {s : Submonoid α} {a₁ b₁ : α} {a₂ b₂ : s}
+
+@[to_additive]
+theorem mk_left_injective (b : s) : Injective fun a => mk a b := fun c d h => by
+  have : Nonempty s := ⟨b⟩ -- porting note: Needed to add this `have`
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/MonoidLocalization.lean:1898]: Do you know why? Did we misport a lemma and add this hypothesis unecessarily?

Comment (YaelDillies) [reply Mathlib/GroupTheory/MonoidLocalization.lean:1898]: No, I cannot find the lemma that uses it. It might be a matter of instance unification, *again*.

Comment (eric-wieser) [reply Mathlib/GroupTheory/MonoidLocalization.lean:1898]: I think we're missing the fact that submonoids are inhabited (with 1)

Comment (eric-wieser) [reply Mathlib/GroupTheory/MonoidLocalization.lean:1898]: In Lean3 this finds `has_one.nonempty`

Comment (YaelDillies) [reply Mathlib/GroupTheory/MonoidLocalization.lean]: No fix in sight, then?

Comment (eric-wieser) [reply Mathlib/GroupTheory/MonoidLocalization.lean]: I think TC search just times out before it finds it

**Comment by eric-wieser** [comment]
bors merge

Thanks!

**Pull request closed without merging**